### DISCUSSION TITLE
Switching homebrew to work without sudo

### DISCUSSION THREAD

**Discussion by EricFriedlander** [Homebrew/discussions#3259] [Everyday usage]
Hello!

My school (where I'm employed) got randsom-wared last last year. As a result, they now have installed "admin-by-request" on all school owned computers and I no longer have sudo access. As I understand it (and I apologize if this is wrong) homebrew can work when you don't have sudo access. However, when I originally installed homebrew, I had sudo access so now when I try to install things I get "<username> is not in the sudoers file. This incident will be reported.". Is there a way to switch homebrew so that it no longer requires sudo access without breaking anything? I can potentially get sudo access temporarily to make the changes if necessary.

Thanks!

**Comment by SMillerDev** [comment]
A new homebrew install only requires sudo when installing homebrew itself. What are you running that gives this error?

**Reply by EricFriedlander** [reply]
Anytime I try to install anything it gives me that error. Here is an example of the output I get when trying to install microsoft edge:

```
brew install --cask microsoft-edge
Running `brew update --preinstall`...
==> Auto-updated Homebrew!
Updated 2 taps (homebrew/core and homebrew/cask).
==> Updated Formulae
Updated 62 formulae.
==> New Casks
aethersx2                  manila                     podman-desktop
ferdium                    oxwu                       protokol
groestlcoin-core           plus42-binary              vieb
jquake                     plus42-decimal             yandex-music-unofficial
==> Updated Casks
Updated 118 casks.
==> Deleted Casks
twitch

==> Downloading https://officecdn-microsoft-com.akamaized.net/pr/03adf619-38c6-4
Already downloaded: /Users/frieeb/Library/Caches/Homebrew/downloads/fb238ef16b6038b107f6a588e9bfa011d489bf251b1f0ade97901c4243a90496--MicrosoftEdge-101.0.1210.39.pkg
==> Installing Cask microsoft-edge
==> Running installer for microsoft-edge; your password may be necessary.
Package installers may write to any location; options such as `--appdir` are ignored.
Password:
Sorry, try again.
Password:
frieeb is not in the sudoers file.  This incident will be reported.
==> Purging files for version 101.0.1210.39 of Cask microsoft-edge
Error: Failure while executing; `/usr/bin/sudo -E -- /usr/bin/env LOGNAME=frieeb USER=frieeb USERNAME=frieeb /usr/sbin/installer -pkg /opt/homebrew/Caskroom/microsoft-edge/101.0.1210.39/MicrosoftEdge-101.0.1210.39.pkg -target / -applyChoiceChangesXML /private/tmp/choices20220511-58874-frwhek.xml` exited with 1. Here's the output:
frieeb is not in the sudoers file.  This incident will be reported.
```

**Reply by gromgit** [reply]
That's the Edge installer (`MicrosoftEdge-101.0.1210.39.pkg`) that's prompting you, not Homebrew, If you downloaded and installed the Edge installer package manually, you'd probably get exactly the same error.

Most casks work by running the upstream installer, so they're similarly constrained if you don't have the sudo access that's sometimes required. If you were installing _formulae_, like `brew install go`, and you got a sudo password prompt, that would be a major surprise, and possibly an indication that someone's hacked your system.

**Reply by EricFriedlander** [reply]
Hmmm. That makes sense. The question I have is that my school uses "admin by request". When I request administrative privileges, I have no problem with the Edge installer. However, the homebrew installation still fails with the same message. Any idea why that could be?

**Reply by gromgit** [reply]
Apologies, was posting before :coffee:. Installing casks _does_ invoke `sudo`, as your error message actually indicated, and my own test confirms:
```
$ brew install --debug --verbose --cask microsoft-edge
[...]
==> Installing artifact of class Cask::Artifact::Pkg
==> Running installer for microsoft-edge; your password may be necessary.
Package installers may write to any location; options such as `--appdir` are ignored.
/usr/bin/sudo -E -- /usr/bin/env LOGNAME=aho USER=aho USERNAME=aho /usr/sbin/installer -pkg /opt/homebrew/Caskroom/microsoft-edge/101.0.1210.39/MicrosoftEdge-101.0.1210.39.pkg -target / -verboseR -applyChoiceChangesXML /opt/homebrew/tmp/choices20220512-58408-9ndoa1.xml
Password:
```
and I see from https://www.adminbyrequest.com/Mac that:
> For security reasons, sudo access is disabled during administrator sessions by default. This can be enabled in the settings or a policy file (see [Mac policies](https://www.adminbyrequest.com/PoliciesMac)). We do not recommend enabling sudo access unless absolutely necessary. Admin By Requests has checks in place to prevent system tampering using sudo, but due to the root-level access, it is impossible to fully protect against tampering using sudo. Consider using the build-in /etc/sudoers file, if only certain commands needs to be run with sudo. The Admin By Request sudo settings will not override normal /etc/sudoers settings.

Since `sudo` is used to invoke the macOS system install tool, and given your environment runs Admin By Request because malware got installed, I doubt your admins would be happy to add that command to `/etc/sudoers`. I'm afraid enabling the required access is something you'll have to discuss with them.

**Reply by EricFriedlander** [reply]
Am I right in understanding that it is possible to set-up homebrew to not use sudo? And if so, how can I switch it without messing things up. Assume that I can get temporary sudo access to make the change...

**Reply by SMillerDev** [reply]
If you only use formula you will not need sudo. But that is not a common usecase for homebrew

**Reply by Kentzo** [reply]
Note that it's a flaw in the current implementation of Homebrew to unconditionally require `sudo` for installer packages (files with the pkg extension like in your case).

I'm trying to address this via https://github.com/Homebrew/brew/pull/14593
